compiler/rustc_ast/src/ast.rs

```diff
@@ -2733,6 +2733,13 @@ impl FnRetTy {
             FnRetTy::Ty(ty) => ty.span,
         }
     }
+
+    pub fn has_ret(&self) -> bool {
```

Comment:
This is very much sketchy 👍 

Ideally don't distinguish between `-> ()` and no return, because in Rust these are precisely equal, but in any case, this should *not* be a helper that other users may accidentally use since it's 99.9% wrong.

Reply:
I don't think closures are relevant enough to spend time on adding support for it, my macro rejects everything but functions.
I don't care about the difference between -> () and the implicit () return. It looks like it worked by coincidence before correctly, but I've added a check for explicit -> () return and also moved the function out of ast.rs

Reply:
Question: does it matter to autodiff then if the return type is a type-alias-to-unit or effectively-unit?

```rs
// e.g.
type Unit = ();
// or 
#[repr(transparent)]
pub struct UnitButFancy { inner: () }
```

Then the user may have written

```
pub fn foo() -> Unit /* or `UnitButFancy` */ {}
```

Would this be a problem?

Reply:
I don't really know how the implementation would behave here.

The current mvp strongly relies on the user not doing "strange" things, I don't see this experiment as anywhere close of handling type aliases or malicious users correctly. Sicne there are still bugs left that users in good faith will run into, I'd just leave this as an unhandled case. This type doesn't have any floats in it so const would be the only reasonable annotation, but I don't think spending time on catching this would be a good investment.

Reply:
That seems fine, probably worth remarking this as a comment.

Reply:
Yep, by now we have the info that we would like to support this both in the autodiff_attrs and the new tests.